src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs

```diff
@@ -210,5 +214,76 @@ public override bool GeneratesPInvoke(MethodDesc method)
 
             return !Marshaller.IsMarshallingRequired(method);
         }
+
+        public override bool TryGetModuleTokenForExternalType(TypeDesc type, out ModuleToken token)
+        {
+            Debug.Assert(!VersionsWithType(type));
+
+            if (_typeRefsInCompilationModuleSet == null)
+            {
+                _typeRefsInCompilationModuleSet = new Dictionary<TypeDesc, ModuleToken>();
+
+                foreach (var module in _compilationModuleSet)
+                {
+                    EcmaModule ecmaModule = (EcmaModule)module;
+                    foreach (var typeRefHandle in ecmaModule.MetadataReader.TypeReferences)
+                    {
+                        TypeDesc typeRef = ecmaModule.GetType(typeRefHandle);
+                        if (!_typeRefsInCompilationModuleSet.ContainsKey(typeRef))
+                        {
+                            _typeRefsInCompilationModuleSet.Add(typeRef, new ModuleToken(ecmaModule, typeRefHandle));
+                        }
+                    }
+                }
+            }
+
+            return _typeRefsInCompilationModuleSet.TryGetValue(type, out token);
+        }
+
+        private void MapExternalMethodsAndFields()
+        {
+            if (_methodRefsInCompilationModuleSet != null || _fieldRefsInCompilationModuleSet != null)
+                return;
+
+            _methodRefsInCompilationModuleSet = new Dictionary<MethodDesc, ModuleToken>();
+            _fieldRefsInCompilationModuleSet = new Dictionary<FieldDesc, ModuleToken>();
+
+            foreach (var module in _compilationModuleSet)
+            {
+                EcmaModule ecmaModule = (EcmaModule)module;
+                foreach (var memberReferenceHandle in ecmaModule.MetadataReader.MemberReferences)
+                {
+                    object fieldOrMethod = ecmaModule.GetObject(memberReferenceHandle);
+                    if (fieldOrMethod is FieldDesc field && !_fieldRefsInCompilationModuleSet.ContainsKey(field))
```

Comment:
Maybe we could use TryAdd instead of ContainsKey/Add - seems like it would do half the amount of lookup.